api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -495,4 +496,89 @@ public Workspace setPublished(
 
     return this.saveWithLastModified(workspace);
   }
+
+  @Override
+  public List<UserRecentWorkspace> getRecentWorkspacesByUser(long userId) {
+    List<UserRecentWorkspace> userRecentWorkspaces =
+        userRecentWorkspaceDao.findByUserIdOrderByLastAccessDateDesc(userId);
+    enforceFirecloudAclsInRecentWorkspaces(userRecentWorkspaces);
+    return userRecentWorkspaceDao.findByUserIdOrderByLastAccessDateDesc(userId);
+  }
+
+  private void enforceFirecloudAclsInRecentWorkspaces(List<UserRecentWorkspace> recentWorkspaces) {
+    // Map recentWorkspaces by id because we'll want to access them like that later when we delete
+    Map<Long, UserRecentWorkspace> recentWorkspacesById =
+        recentWorkspaces.stream()
+            .collect(Collectors.toMap(UserRecentWorkspace::getWorkspaceId, rw -> rw));
+
+    // Grab the dbWorkspaces.
+    List<Workspace> dbWorkspaces =
+        workspaceDao.findAllByWorkspaceIdIn(
+            recentWorkspaces.stream()
+                .map(UserRecentWorkspace::getWorkspaceId)
+                .collect(Collectors.toList()));
+
```

Comment:
Should you take some action if you can't find all the workspace you were looking for? I.e. should we maybe also remove from the recent access table all workspaces (for all users) that aren't in the DB anymore? There a cron job or cascade that's supposed to handle that, so maybe log an error or warning in this case (i.e. this is a referential integrity problem).

Reply:
I don't know about a cron job, but there's certainly not a cascade that's supposed to handle this. We don't have a DB table for who has access to what DB. FireCloud is the source of truth for all workspace permissions.

I don't think this is a referential integrity problem per se. I do delete the referenced `user_recent_workspace` entry when a `workspace` or `user` is deleted. The only stale data in here is a recent workspace that had been shared with the user that is unshared. That's what I'm handling here. If I were to change it, I would handle it at the point where it's being unshared, but we don't have control over all those points so I'd still need to keep this function.